### PR TITLE
#8: Provider chain

### LINKED ISSUE #8: Should use AWS default credential provider chain
Hi,
it should be possible to use this package without defining the `AWS_KEY` and `AWS_SECRET` env variable. 

It's recommended to use the default credential provider chain as describe [here](https://docs.aws.amazon.com/sdk-for-php/v3/developer-guide/guide_credentials.html).

So this should be unnecessary:
https://github.com/lizards-and-pumpkins/lib-queue-backend-amazon-sqs/blob/ef4897ad16755abc5647dd093679aaffe7f90e98/src/SqsFactory.php#L55-L58

We need to be able to assume an IAM role. The new env variables would be `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY`

Thanks :)

### PR DESCRIPTION
**Pull request by Schrank**
Closes #8

### CODE REVIEW

File: src/SqsFactory.php [hunk 1]
```diff
@@ -36,7 +38,11 @@ private function getEventQueueUrl(): string
         /** @noinspection PhpUndefinedMethodInspection */
         $configReader = $this->getMasterFactory()->createConfigReader();
 
-        return $configReader->get('AWS_SQS_EVENT_QUEUE_URL');
+        try {
+            return $configReader->get('AWS_SQS_EVENT_QUEUE_URL');
+        } catch (\TypeError $e) {
+            throw new MissingConfigurationException('Please pass AWS_SQS_EVENT_QUEUE_URL as env variable.');
+        }
```
Comment (LlZARD) [inline src/SqsFactory.php]: What is the purpose here? `EnvironmentConfigReader::get()` is already throwing an exception.

File: src/SqsFactory.php [hunk 2]
```diff
@@ -45,21 +51,36 @@ private function getCommandQueueUrl(): string
         /** @noinspection PhpUndefinedMethodInspection */
         $configReader = $this->getMasterFactory()->createConfigReader();
 
-        return $configReader->get('AWS_SQS_COMMAND_QUEUE_URL');
+        try {
+            return $configReader->get('AWS_SQS_COMMAND_QUEUE_URL');
+        } catch (\TypeError $e) {
+            throw new MissingConfigurationException('Please pass AWS_SQS_COMMAND_QUEUE_URL as env variable.');
+        }
```
Comment (LlZARD) [inline src/SqsFactory.php]: Same here.

File: src/SqsFactory.php [hunk 3]
```diff
@@ -45,21 +51,36 @@ private function getCommandQueueUrl(): string
...
+        try {
+            $credentials = [
+                'key' => $this->getAwsKey(),
+                'secret' => $this->getAwsSecret(),
+            ];
+        } catch (EnvironmentConfigKeyIsNotSetException $e) {
+            // Intended, credentials can be provided via default credential provider chain
+            //see https://github.com/lizards-and-pumpkins/lib-queue-backend-amazon-sqs/issues/8
+        } catch (\TypeError $e) {
+            // only happens on unit test
```
Comment (LlZARD) [inline src/SqsFactory.php]: This smells a lot. Why unit test is throwing an exception?

File: src/SqsFactory.php [hunk 4]
```diff
@@ -45,21 +51,36 @@ private function getCommandQueueUrl(): string
...
+        }
+        $credentials = [];
+        try {
+            $credentials = [
+                'key' => $this->getAwsKey(),
+                'secret' => $this->getAwsSecret(),
+            ];
+        } catch (EnvironmentConfigKeyIsNotSetException $e) {
+            // Intended, credentials can be provided via default credential provider chain
+            //see https://github.com/lizards-and-pumpkins/lib-queue-backend-amazon-sqs/issues/8
```
Comment (LlZARD) [inline src/SqsFactory.php]: This is dirty also. Why not use `ConfigReader::has()` to check if config exists?

File: src/SqsFactory.php [hunk 5]
```diff
@@ -45,21 +51,36 @@ private function getCommandQueueUrl(): string
...
+        } catch (\TypeError $e) {
+            // only happens on unit test
         }
 
+        $this->sqsClient = SqsClient::factory([
+            'credentials' => $credentials,
+            'region' => $this->getAwsRegion(),
+        ]);
```
Comment (LlZARD) [inline src/SqsFactory.php]: I would also extract it into a `getAwsCredentials` utility method.

File: src/SqsFactory.php [hunk 6]
```diff
@@ -68,8 +89,11 @@ private function getAwsRegion(): string
         /** @var ConfigReader $configReader */
         /** @noinspection PhpUndefinedMethodInspection */
         $configReader = $this->getMasterFactory()->createConfigReader();
-
-        return $configReader->get('AWS_REGION');
+        try {
+            return $configReader->get('AWS_REGION');
+        } catch (\TypeError $e) {
+            throw new MissingConfigurationException('Please pass AWS_REGION as env variable.');
+        }
```
Comment (LlZARD) [inline src/SqsFactory.php]: Same as above.

**Comment by Schrank** [comment]
@LlZARD I hope you got everything.

File: src/Exception/MissingConfigurationException.php [hunk 7]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+declare(strict_types=1);
+
+namespace LizardsAndPumpkins\Messaging\Queue\Sqs\Exception;
+
+class MissingConfigurationException extends \RuntimeException
+{
+
+}
```
Comment (LlZARD) [inline src/Exception/MissingConfigurationException.php]: I guess this class can now go away.

File: src/SqsFactory.php [hunk 8]
```diff
@@ -51,11 +51,10 @@ private function getCommandQueueUrl(): string
     private function getSqsClient(): SqsClient
     {
         if (! $this->sqsClient) {
+            $credentials = $this->getAwsCredentials();
```
Comment (LlZARD) [inline src/SqsFactory.php]: Superfluous variable. Please inline this call.

File: src/SqsFactory.php [hunk 9]
```diff
@@ -72,21 +71,19 @@ private function getAwsRegion(): string
         return $configReader->get('AWS_REGION');
     }
 
-    private function getAwsKey(): string
+    private function getAwsCredentials(): array
```
Comment (LlZARD) [inline src/SqsFactory.php]: Please add a doc block.

File: tests/Unit/Suites/LizardsAndPumpkins/Messaging/Queue/Sqs/SqsFactoryTest.php [hunk 10]
```diff
@@ -59,11 +49,48 @@ public function testImplementsMessageQueueFactoryInterface(): void
...
+            ['AWS_KEY', false],
+            ['AWS_SECRET', false],
+        ]);
+
+        $this->masterFactoryMock->method('createConfigReader')->willReturn($configReader);
+
         $this->assertInstanceOf(Queue::class, $this->sqsFactory->createCommandMessageQueue());
     }
+
+    private function prepareConfigReader(): void
```
Comment (LlZARD) [inline tests/Unit/Suites/LizardsAndPumpkins/Messaging/Queue/Sqs/SqsFactoryTest.php]: Do we still support PHP 7.0.x for this repository?

Comment (Schrank) [reply tests/Unit/Suites/LizardsAndPumpkins/Messaging/Queue/Sqs/SqsFactoryTest.php]: No we don't :-) I decline to support unsupported php versions.

Comment (LlZARD) [reply tests/Unit/Suites/LizardsAndPumpkins/Messaging/Queue/Sqs/SqsFactoryTest.php]: Ah, I had to look into `composer.json`. Sorry for a dumb comment. 🙂

Comment (Schrank) [reply tests/Unit/Suites/LizardsAndPumpkins/Messaging/Queue/Sqs/SqsFactoryTest.php]: I documented it in composer.json? Very good past @Schrank :D

**Pull request merged**